### PR TITLE
rgw: s3 bucket logging/server access logging

### PR DESCRIPTION
**Pull request by mikulely**
This is initial impl of s3 bucket logging/server access logging (
http://docs.aws.amazon.com/AmazonS3/latest/dev/ServerLogs.html )

We've discussed this on ML before (
https://www.spinics.net/lists/ceph-devel/msg35681.html )

Usage doc can be found at https://github.com/mikulely/ceph/commit/dd5213544c12f4887186cd21413930420e2eea84

I've also writen a shell script as runbook to demonstrate BL usage can be found at
https://gist.github.com/mikulely/cb6b245c60fda4c1546aaa761c2e5615

We've rebased this functionality form Jewel to Master, and added test
cases to s3-tests in the following pr https://github.com/ceph/s3-tests/pull/161

~~Current impl lack of Log Delivery Group ACL support
(http://docs.aws.amazon.com/AmazonS3/latest/dev/enable-logging-programming.html), 
to someone who may concern.~~

~~potential issue/ future work~~

~~+ race condition when opslog add to and bl deliver from same opslog rdos-obj~~
~~+ cleanup those opslog rados-obj which opslog is enable but bl is disable~~
  
~~Those issue need future work to refactor current ops log impl.~~

~~Maybe we can rename opslog rados-obj when delivering, cause rados
object are located by name, rename means extra data copy.~~

~~We propose that we can organize opslogs into rados namespace,
namespace can be named by date(like 2017-04-47),and bl_deliver only
process namespace with older date.This process fit our deliver
scheduler(provide bucket logging for yesterday) and also avoiding race
condition. cleanup unused opslog also is very easy，just remove remaining
opslog objs in the namespace after delivery.~~

CC: @mattbenjamin @yehudasa @robbat2 (Plz take a test if you're still interested.)

Edited:

2017-06-23: strike through outdated description

### CODE REVIEW

**Comment by robbat2** [comment]
@mikulely does this correctly handle load-balanced multiple RGW instances (eg haproxy in front of RGW)?  I haven't read the entire series to see if it would, but wanted to ask anyway.

**Comment by mattbenjamin** [comment]
@mikulely from RGW bug scrub:  do you have unit test support, or can provide?

**Comment by mikulely** [comment]
@robbat2 what does `load-balanced multiple RGW instances` mean? Let's clarify that the opslog record procedure is achieved by `rgw_enble_ops_log` impl , this pr only upload them to user bucket.

@mattbenjamin there's no unit test at function level, but we do write [a bunch of python scripts](https://github.com/ZVampirEM77/rgw_bl_auto_test_script)  for functionality test.

**Comment by robbat2** [comment]
@mikulely The upload to user bucket, does it correctly take into account that multiple RGWs may be concurrently writing different entries in the ops log? Eg, during a multipart upload, individual parts may be concurrently uploaded to different RGWs.

**Comment by mikulely** [comment]
@robbat2 1. There‘s lock mechanism to guard only one rgw instance can handle certain bucket's opslog(copy from object lifecycle management  impl )2. Every time rgw upload opslog to a unique rgw-object,the rgw-object name is generated by https://github.com/ceph/ceph/pull/14841/files#diff-ff97b7e16bb0ec2eef20f8068c88726cR158

**Comment by mikulely** [comment]
rebased with master

**Comment by mikulely** [comment]
Repushed and add LDG ACL support

File: src/common/legacy_config_opts.h [hunk 1]
```diff
@@ -1459,12 +1459,41 @@ OPTION(rgw_override_bucket_index_max_shards, OPT_U32, 0)
...
+OPTION(rgw_enable_bl_threads, OPT_BOOL, true)
+
+OPTION(rgw_bl_max_objs, OPT_INT, 32)
+
+OPTION(rgw_bl_lock_max_time, OPT_INT, 60)  // total run time for a single bl processor work
+OPTION(rgw_bl_work_time, OPT_STR, "00:00-06:00") //job process bl at 00:00-06:00s
+/**
+ * bl thread need to upload opslog by sending http put requests to itself
+
+ * We need to support FCGI frontend which we can't find listening port
```
Comment (amitkumar50) [inline src/common/legacy_config_opts.h]: @mikulely It would be good if we can use `can not` instead of `can't`

File: src/rgw/rgw_acl_s3.cc [hunk 2]
```diff
@@ -18,9 +18,11 @@ using namespace std;
 
 #define RGW_URI_ALL_USERS	"http://acs.amazonaws.com/groups/global/AllUsers"
 #define RGW_URI_AUTH_USERS	"http://acs.amazonaws.com/groups/global/AuthenticatedUsers"
+#define RGW_URI_LOG_DELIVERY    "http://acs.amazonaws.com/groups/s3/LogDelivery" //TODO cleanup redundant define
```
Comment (amitkumar50) [inline src/rgw/rgw_acl_s3.cc]: @mikulely http://acs.amazonaws.com/groups/s3/LogDelivery
This site can’t be reached

Comment (robbat2) [inline src/rgw/rgw_acl_s3.cc]: It's NOT a website, it's a magic ACL string in URL format. You'll see it already hardcoded in Boto.

Comment (amitkumar50) [inline src/rgw/rgw_acl_s3.cc]: @robbat2 ok. Thanks for Confirmation

File: src/rgw/rgw_bl.cc [hunk 3]
```diff
@@ -0,0 +1,984 @@
...
+      ("WRITE_ACP", RGW_PERM_WRITE_ACP)
+      ("FULL_CONTROL", RGW_PERM_FULL_CONTROL);
+
+void *RGWBL::BLWorker::entry() {
+  utime_t end = ceph_clock_now();
+  do {
+    utime_t start = ceph_clock_now();
+    if (should_work(start) &&
+        (start.sec() - end.sec() >= bl->deliver_interval)) {
+      dout(5) << "bucket logging deliver: start" << dendl;
```
Comment (yehudasa) [inline src/rgw/rgw_bl.cc]: should use ldout

File: src/rgw/rgw_bl.cc [hunk 4]
```diff
@@ -0,0 +1,984 @@
...
+      dout(5) << "bucket logging deliver: stop" << dendl;
+    }
+    if (bl->going_down())
+      break;
+
+    end = ceph_clock_now();
+    int secs = schedule_next_start_time(end);
+    time_t next_time = end + secs;
+    char buf[30];
+    char *nt = ctime_r(&next_time, buf);
```
Comment (yehudasa) [inline src/rgw/rgw_bl.cc]: there's `utime_t::asctime()` that can be used

File: src/rgw/rgw_bl.cc [hunk 5]
```diff
@@ -0,0 +1,984 @@
...
+        dout(0) << "ERROR: bucket logging process() err=" << r << dendl;
+      }
+      dout(5) << "bucket logging deliver: stop" << dendl;
+    }
+    if (bl->going_down())
+      break;
+
+    end = ceph_clock_now();
+    int secs = schedule_next_start_time(end);
+    time_t next_time = end + secs;
```
Comment (yehudasa) [inline src/rgw/rgw_bl.cc]: why convert to `time_t`?

File: src/rgw/rgw_bl.cc [hunk 6]
```diff
@@ -0,0 +1,984 @@
...
+                 << referrer << row_separator                                  // Referrer
+                 << user_agent << row_separator                                // User-Agent
+                 << oversion_id << row_separator                               // Version Id
+                 << newliner;
+
+  buffer->append(pending_column.str());
+}
+
+int RGWBL::bucket_bl_upload(bufferlist* opslog_buffer, rgw_obj obj,
+			    map<string, bufferlist> tobject_attrs)
```
Comment (yehudasa) [inline src/rgw/rgw_bl.cc]: Need to rewrite this method to create an object directly instead of sending a message to ourselves

**Comment by mikulely** [comment]
@yehudasa I've tested with AWS S3 bucket logging, the behavior of S3 would record put log file operation as a record in the log file itself. 

PS: Will  rebase against to master next week after fix the compatibility issue of operation field(https://github.com/ceph/ceph/pull/16984).

**Comment by yehudasa** [comment]
@mikulely that is fine, but we can achieve that without needing to send ourselves messages.

**Comment by mikulely** [comment]
@yehudasa do you mean issue a  `RGWPutObj_ObjStore_S3` op manually?

File: src/common/legacy_config_opts.h [hunk 7]
```diff
@@ -1459,12 +1459,41 @@ OPTION(rgw_override_bucket_index_max_shards, OPT_U32, 0)
...
+ * whether or not the bl threads should be started
+ */
+OPTION(rgw_enable_bl_threads, OPT_BOOL, true)
+
+OPTION(rgw_bl_max_objs, OPT_INT, 32)
+
+OPTION(rgw_bl_lock_max_time, OPT_INT, 60)  // total run time for a single bl processor work
+OPTION(rgw_bl_work_time, OPT_STR, "00:00-06:00") //job process bl at 00:00-06:00s
+/**
+ * bl thread need to upload opslog by sending http put requests to itself
```
[the remark this answers is not in the document]
Comment (mikulely) [reply src/common/legacy_config_opts.h]: @yehudasa Using HTTP will let us send opslogs into bucket that created in other zonegroup, here's a special case for send opslogs into current zonegroup.

File: src/rgw/rgw_admin.cc [hunk 8]
```diff
@@ -157,6 +158,8 @@ void _usage()
   cout << "  gc process                 manually process garbage\n";
   cout << "  lc list                    list all bucket lifecycle progress\n";
   cout << "  lc process                 manually process lifecycle\n";
+  cout << "  bl list                    list all bucket logging deliver progress\n";
+  cout << "  bl process                 deliver bucket logging manually\n";
```
[the remark this answers is not in the document]
Comment (mikulely) [reply src/rgw/rgw_admin.cc]: I'm fine with both, if we're going to use `bucket log`, maybe we need also change `lc` to `bucket lc` for consistency.

File: src/rgw/rgw_admin.cc [hunk 9]
```diff
@@ -2476,6 +2491,8 @@ int main(int argc, const char **argv)
       admin_specified = true;
     } else if (ceph_argparse_binary_flag(args, i, &system, NULL, "--system", (char*)NULL)) {
       system_specified = true;
+    } else if (ceph_argparse_binary_flag(args, i, &bl_deliver, NULL, "--bl_deliver", (char*)NULL)) {
```
[the remark this answers is not in the document]
Comment (mikulely) [reply src/rgw/rgw_admin.cc]: addressed

File: src/rgw/rgw_bl.cc (hunk 3 above)
Comment (mikulely) [reply src/rgw/rgw_bl.cc]: addressed.

File: src/rgw/rgw_bl.cc (hunk 4 above)
Comment (mikulely) [reply src/rgw/rgw_bl.cc]: ok

File: src/rgw/rgw_bl.cc (hunk 5 above)
Comment (mikulely) [reply src/rgw/rgw_bl.cc]: just copied from `rgw_lc.cc`.

File: src/common/legacy_config_opts.h (hunk 1 above)
Comment (mikulely) [reply src/common/legacy_config_opts.h]: will address when adapt bl to new config interface.

**Comment by mikulely** [comment]
@yehudasa rebased against master：

+ fix compiler errno
+ merge operation field support ( pr https://github.com/ceph/ceph/pull/17754)

Plz help to review.

**Comment by mattbenjamin** [comment]
unstale me my man

File: src/rgw/rgw_log.h [hunk 10]
```diff
@@ -108,6 +116,11 @@ struct rgw_log_entry {
     if (struct_v >= 9) {
       ::decode(x_headers, p);
     }
+    if (struct_v >= 10) {
+      ::decode(prot_flags, p);
```
Comment (chuang-he) [inline src/rgw/rgw_log.h]: ::decode(request_id, p); ?

**Pull request closed without merging**